app/services/users/username_generator.rb

```diff
@@ -7,46 +7,55 @@
 # @todo Extract username validation in separate class
 module Users
   class UsernameGenerator
+    attr_reader :usernames
+
     def self.call(...)
       new(...).call
     end
 
-    # @param list [Array<String>] a list of usernames
-    def initialize(list = [])
-      @list = list
+    # @param usernames [Array<String>] a list of usernames
+    def initialize(usernames = [], detector: CrossModelSlug, generator: nil)
+      @detector = detector
+      @generator = generator || method(:random_username)
+      @usernames = usernames
     end
 
     def call
-      from_list(modified_list) || from_list(list_with_suffix) || from_list(Array.new(3) { random_letters })
+      first_available_from(normalized_usernames) ||
+        first_available_from(suffixed_usernames) ||
+        first_available_from(random_usernames)
     end
 
-    private
+    def normalized_usernames
+      @normalized_usernames ||= filtered_usernames.map { |s| s.downcase.gsub(/[^0-9a-z_]/i, "").delete(" ") }
+    end
 
-    def from_list(list)
-      list.detect { |username| !username_exists?(username) }
+    def filtered_usernames
+      @filtered_usernames ||= usernames.select { |s| s.is_a?(String) && s.present? }
     end
 
-    def username_exists?(username)
-      User.exists?(username: username) ||
-        Organization.exists?(slug: username) ||
-        Page.exists?(slug: username) ||
-        Podcast.exists?(slug: username)
```

Comment:
As discussed [over here](https://github.com/forem/forem/pull/19252/files#r1147814236), refactoring the name generator to use the new cross-model-unique-slug gizmo.

Reply:
Shouldn't we apply the reserved list validator?

Reply:
Good point! I mean, generator generates before validation, but the generator probably wants to move onto the next attempt if it randomly generates a `ReservedWord`, rather than block user create. I believe the extracted `CrossModelSlug` class is going to be useful elsewhere, so I added it [there](https://github.com/forem/forem/pull/19263/commits/5b6e2233d17224c2505ba82fc8f4d2fefb6e4438) rather than add it to the generator class.